ipam/allocator.go

```diff
@@ -0,0 +1,251 @@
+package ipam
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	lg "github.com/zettio/weave/common"
+	"github.com/zettio/weave/ipam/ring"
+	"github.com/zettio/weave/ipam/space"
+	"github.com/zettio/weave/ipam/utils"
+	"github.com/zettio/weave/router"
+)
+
+const (
+	tombstoneTimeout = 14 * 24 * time.Hour
+)
+
+// Kinds of message we can unicast to other peers
+const (
+	msgSpaceRequest = iota
+	msgLeaderElected
+	msgRingUpdate
+)
+
+type pendingAllocation struct {
+	resultChan chan<- net.IP
+	Ident      string
+}
+
+// Allocator brings together Ring and space.Set, and does the nessecary plumbing.
+type Allocator struct {
+	actionChan    chan<- func()
+	ourName       router.PeerName
+	universeStart net.IP
+	universeSize  uint32
+	universeLen   int        // length of network prefix (e.g. 24 for a /24 network)
+	ring          *ring.Ring // it's for you!
+	spaceSet      space.Set
+	owned         map[string][]net.IP // who owns what address, indexed by container-ID
+	pending       []pendingAllocation
+	claims        claimList
+	gossip        router.Gossip
+	shuttingDown  bool
+}
+
+// NewAllocator creats and initialises a new Allocator
+func NewAllocator(ourName router.PeerName, universeCIDR string) (*Allocator, error) {
+	_, universeNet, err := net.ParseCIDR(universeCIDR)
+	if err != nil {
+		return nil, err
+	}
+	if universeNet.IP.To4() == nil {
+		return nil, errors.New("Non-IPv4 address not supported")
+	}
+	// Get the size of the network from the mask
+	ones, bits := universeNet.Mask.Size()
+	var universeSize uint32 = 1 << uint(bits-ones)
+	if universeSize < 4 {
+		return nil, errors.New("Allocation universe too small")
+	}
+	alloc := &Allocator{
+		ourName:       ourName,
+		universeStart: universeNet.IP,
+		universeSize:  universeSize,
+		universeLen:   ones,
+		ring:          ring.New(utils.Add(universeNet.IP, 1), utils.Add(universeNet.IP, universeSize-1), ourName),
+		owned:         make(map[string][]net.IP),
+	}
+	return alloc, nil
+}
+
+func (alloc *Allocator) string() string {
```

Comment:
ah yes. Once again I got confused by which file these methods are in. Yes, inline it. It's short enough.